src/main/java/com/atlassian/db/replica/internal/DecisionAwareReference.java

```diff
@@ -0,0 +1,44 @@
+package com.atlassian.db.replica.internal;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class DecisionAwareReference<T> {
+    private final LazyReference<T> lazyReference = new InnerLazyReference<>(this);
+    private final AtomicReference<RouteDecisionBuilder> decisionBuilder = new AtomicReference<>();
+
+    public abstract T create() throws Exception;
+
+    public boolean isInitialized() {
+        return lazyReference.isInitialized();
+    }
+
+    public T get(RouteDecisionBuilder decisionBuilder) {
+        if (!isInitialized()) {
+            this.decisionBuilder.compareAndSet(null, decisionBuilder);
+        }
+        return lazyReference.get();
+    }
+
+    public void reset() {
+        this.lazyReference.reset();
+        this.decisionBuilder.set(null);
+    }
+
+    public RouteDecisionBuilder getDecisionBuilder() {
+        return decisionBuilder.get();
```

Comment:
Returns null if `!isInitialized`. Go `Optional`?

Reply:
So now we have:
```java
if (this.decisionBuilder.get() == null) {
    throw new IllegalStateException("The decision builder is not initialized");
}
return decisionBuilder.get();
```
So still a runtime error. I suggested a compile-time error. I'll include it in my PR. If we want this to be a runtime error, let me know.

Reply:
Hmm, ok, there's no sensible handling of an empty (missing) decision. So right now it relies on `get` being called earlier (which is safe to assume within `create`). I'll think if I can simplify this assumption

Reply:
I made 2 attempts to make it a compile-time error. It led to a broad refactoring. IMO the PR is already complex, and the internals can be refactored later.

Reply:
Fair enough. It's internal anyway 👍